### PR TITLE
[dtensor] introduce experimental dmap

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103686

This PR introduces `dmap`, this allows user to write function with local
shard view inside the function, i.e. write manual collectives. the dmap
can work together with DTensor operations.

i.e. We can do some DTensor computation first, then call a dmap function
to write some manual collectives, then use the results to do some
additional DTensor computations.

Given that we already have APIs to explicitly convert torch.Tensor to/from
DTensor, this API just serves as a convenient tool to write simple funcs
that works seamlessly with DTensor.

caveat: manual collectives always need to couple with autograd.Function so
that it works for backward. Ideally we should just make our functional collectives
to have gradient formulas, but that might be a bit tricky to implement.

### CODE REVIEW

**Comment by ezyang** [comment]
Neat! I wonder if it really only just `to_local` and `from_local`?

Can we do the full example at https://jax.readthedocs.io/en/latest/jep/14273-shard-map.html#a-realistic-transformer-example in the test? It would help give me more confidence that this really is equivalent to shmap.

> it turns out not quite hard to enable similar things (not exactly the same behavior as “shamp” as we are in two different programming paradigm, jax is single process, we are doing multiprocess)

It's not clear to me what the distinction here is? If we're SPMD, you should be able to generate equivalent APIs no?

**Comment by wanchaol** [comment]
Another simple fix is to make the fake_pg capable of doing some simple validation logic in allgather, https://github.com/pytorch/pytorch/pull/104213

Agreed we can also rely on MTPG for correctness, but we would need a way to make device mesh be constructed under fake pg so that we can initialize the device mesh in a single device/rank.

File: torch/distributed/_tensor/experimental/dmap.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    """
+    dmap is distributed map, ``dmap(func)`` returns a function that maps the `func` over the
+    dtensor inputs to their local shards. This let user write functions in a local (per-device)
+    mode, instead of the DTensor "global" mode. i.e. we can write manual collectives
+    inside and mix the function call with DTensor operations.
+
+    Args:
+        func (Callable): A Python function that takes one or more arguments.
+            Must return one or more Tensors.
+        mesh (:class:`DeviceMesh`): DeviceMesh that encodes the mesh topology
```
Comment (joshlk) [inline torch/distributed/_tensor/experimental/dmap.py:56]: Can `mesh` be inferred from the DTensor inputs?

**Review by joshlk — COMMENTED** [review]
Hi, I'm currently looking for a tool to deal with operations that don't shard propagate and this seems to do the trick. Ideally to make it easier when maintaining a DTensor and Tensor code path in a module: it would also accept Tensor (non-DTensor) inputs and if all inputs are Tensors then the output should also be a Tensor.

**Review by ezyang — APPROVED** [review]

File: torch/distributed/_tensor/experimental/dmap.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/_tensor/experimental/dmap.py:56]: good point yeah, I think we can remove this mesh and totally infer this from dtensor inputs
